### ISSUE TITLE
DH CHecklist link issue...

### ISSUE THREAD

**Issue by bds002** [magfest/ubersystem#1382]
So in the department head Emails....
We have the phrase:  You can review the complete checklist of all information we'll eventually need from you at https://classic.uber.magfest.org/uber/dept_checklist

The link needs to be update to:
https://classic.uber.magfest.org/uber/dept_checklist/index

I am getting reports that without the /index that some people are actually getting logged out instead of seeing the /index page.   There are multiple emails to fix, but I assume they are pulling from 1 source script for the classic instance.

**Comment by earl7399** [comment]
#1 - I get the checklist
#2 - Is the webserver misconfigured somehow?  That redirect should be
automatic, right?

--
Robert Earl <r.a.earl@gmail.com>
FAX: 1-440-919-5086

**Comment by kitsuta** [comment]
Yes, this is not an issue with the email - that URL should work. We depend on that redirect on a regular basis, so we need to fix the underlying issue.

**Comment by earl7399** [comment]
This feels very server-ish to me - @binary1230 ??

**Comment by EliAndrewC** [comment]
Oh, I know what's happening:
- someone clicks the link in the email
- they are not logged in, and therefore are redirected to the login form
- they log in with their credentials and then are taken to the default homepage
- they're now like, "WTF I don't see a dept head checklist"

I had one department head mention this to me last year as they reason why they never filled out the checklist.

The real fix here is that if someone makes a ``GET`` request and isn't logged in, when we redirect them to the login form we should save what page they were requesting so that after they log in they're taken back to that same page.  Most modern websites do this nowadays :/

**Comment by earl7399** [comment]
This is likely beyond my ability to get done quickly, so I'll stay out of
the way.

--
Robert Earl <r.a.earl@gmail.com>
FAX: 1-440-919-5086

**Comment by bds002** [comment]
Bob, I actually think this is an easy fix and got blown way overboard reading the thread,

Looking at the link the ends:  uber/dept_checklist 

What we are really telling the browser to do is to go to the uber directory and find the file named dept_checklist ... most browsers are smart enough to read that there is no such file, so the browsers clients adds the  trailing slash and default looks for index or index.htm

It almost all of our links and emails etc etc we send the user directly to the page that we want them to see.  (Example: preregistration/confirm?id=83bxxxxxxxxx)

What I would like us to do is the same for the DH Link.
We need to send people directly to the page we want them to see,
Which means adding a trailing slash and the index page name afterwards.
https://classic.uber.magfest.org/uber/dept_checklist/index

To address Eli's portion, I watched the end user log into his Admin account and saw that his browser kicked him out when he tried the link.

**Comment by earl7399** [comment]
​​
To summarize what I just read:
You watched a user:
1 - Log into their admin account in a browser.
2 - Click on the link in the e-mail
3 - NOT end up on https://classic.uber.magfest.org/uber/dept_checklist/overview?message=The%20checklist%20is%20for%20department%20heads%20with%20exactly%20one%20department

Is that true?  Because at step #3, that URL is exactly where we ended up.  The browser is NOT where the work happens to turn "folder/" into "folder/whatever" - that happens at the server, AFAIK.  Which means if is isn't working, it's still a server problem.

Since the link you advocate changing to isn't actually a page on the site, it would still require the server to properly handle what it isn't handling now.  And I'm more than happy to be corrected by @binary1230 / @EliAndrewC / @kitsuta on this, but it still seems to be a server-side issue.

**Comment by bds002** [comment]
Correct the trailing slash and redirect never appeared on his browser.

**Comment by bds002** [comment]
The the trailing slash never appeared ...  
https://classic.uber.magfest.org/uber/dept_checklist/

**Comment by EliAndrewC** [comment]
Hrm, interesting.  That's definitely a problem and is certainly unexpected and needs to be fixed.  We should still do the login thing, but this is definitely a separate bug.

I didn't get home from work until after 11pm last night and tonight is game night, so the soonest I can look into this myself is tomorrow evening, but I'll do that then if no one gets to it first.

**Comment by kitsuta** [comment]
>Looking at the link the ends: uber/dept_checklist
>
>What we are really telling the browser to do is to go to the uber directory and find the file named dept_checklist ... most browsers are smart enough to read that there is no such file, so the browsers clients adds the trailing slash and default looks for index or index.htm

CherryPy is supposed to handle this for us - there really shouldn't be an issue here.

I'm not saying we can't just edit the URL to fix the immediate problem, but even if we do that we need to look into why this is happening. It also isn't universal, as clicking on that link (the one without the trailing slash) doesn't log me out.

Actually, is this reproducible at all, even for the user? It's possible they clicked the link right when we restarted the server, or something.

**Comment by thaeli** [comment]
Is this occurring in the "user is already logged in" condition, or only in "user is not logged in", or both? How about current vs stale vs no session cookies?

**Comment by earl7399** [comment]
It is not occurring for me, and I'm in the "already logged in" situation.

So I switched to a browser I haven't used for a while.  Clicked the link, was redirected to the login page as expected. Logged in, left at the main menu/screen (this is the 'not saving the intended destination during login' issue that @EliAndrewC mentions).  I then click the link again and am taken to the intended URL (https://classic.uber.magfest.org/uber/dept_checklist/overview?message=The%20checklist%20is%20for%20department%20heads%20with%20exactly%20one%20department)

**Comment by binary1230** [comment]
FYI: browser has no concept of / or /index.html or anything like that.  That's all on the server.

as far as the browser is concerned, it just gets http://{hostname}/{querystring}, and querystring is interpreted however it is via the web server.

for the login-remembers-url feature, in cherrypy we should be able to save that querystring and pass it to the next page.  there's some stuff in the HTTPRedirect code that does this kind of thing.
